### PR TITLE
test: autogenerate bash completion

### LINKED ISSUE #17289: contrib: Autogenerate bash completion
We have a bash completion file, which needs to be updated manually: https://github.com/bitcoin/bitcoin/pull/17282#issuecomment-547020006

However, given our argparse module and RPCHelpMan, I believe it should be possible to autogenerate it with maybe a python script.

If it should be necessary to run a bitcoind instance as an http server for rpc request to auto-generate the bash completion, this could be implemented as a functional test (written in python).

### PR DESCRIPTION
**Pull request by suhailsaqan**
Added a functional test which parses all the RPC help commands then automatically generates the bitcoin-cli bash-completion file and makes sure that the original file matches the newly generated one.

In order to get the RPC help commands in the correct format needed for the test, I added the "format" RPC command.

Test using `python3 test/functional/tool_cli_completion.py --overwrite`

### CODE REVIEW

**Comment by fanquake** [comment]
Thanks for picking this up. Did you re-use any code from #18606? A lot of the changes are very similar to the changes there, except for different whitespace, class names, copyright dates etc. If so, cherry-picking so that the original author is preserved is appropriate.

**Comment by laanwj** [comment]
Concept ACK on auto-generating what can be auto-generated. 

I'm not entirely sure about the approach of making this a functional test. It's clever, but also different from how other auto-generated files (the manual pages, the example bitcoin.conf introduced in #22235) are handled in the release process.

On the other hand by adding the check this forces the file on the branch to always be up-to-date so it doesn't need to be in the release process. That's also something.

**Comment by suhailsaqan** [comment]
> Concept ACK on auto-generating what can be auto-generated.
> 
> I'm not entirely sure about the approach of making this a functional test. It's clever, but also different from how other auto-generated files (the manual pages, the example bitcoin.conf introduced in #22235) are handled in the release process.
> 
> On the other hand by adding the check this forces the file on the branch to always be up-to-date so it doesn't need to be in the release process. That's also something.

Look at https://github.com/bitcoin/bitcoin/issues/17289#issue-513520983

**Comment by MarcoFalke** [comment]
Yeah, I wrote that comment before bitcoin.conf moved over to auto-generation. A functional test is one way to achieve this, but this can also be achieved by an auto-generation script at release time.

**Comment by brunoerg** [comment]
From linter:
```
File "test/functional/tool_cli_completion.py" contains a shebang line, but has the file permission 644 instead of the expected executable permission 755. Do "chmod 755 test/functional/tool_cli_completion.py"
```

**Comment by suhailsaqan** [comment]
> I mean you can argue this is a better solution, and that we should move to this for all three. (but for the manual pages it's difficult as they contain a version number as well as the current month)

Thanks for the review @laanwj! Just wondering, do the three of them have to be done the same way? I don't know how I would do it the same way it was done in https://github.com/bitcoin/bitcoin/pull/22235.

**Comment by suhailsaqan** [comment]
Squashed commits.

**Comment by laanwj** [comment]
> do the three of them have to be done the same way? 

Not necessarily, but some structure/standardization around release processes makes it harder to forget any steps, then if everything needs to be done in a different way.

**Comment by luke-jr** [comment]
>On the other hand by adding the check this forces the file on the branch to always be up-to-date so it doesn't need to be in the release process. That's also something.

If we're not careful, it could also increase the rate of diff conflicts (and/or silent merge bugs?).

**Review by hernanmarino — COMMENTED** [review]
Hi everyone, I have a questions / suggestion:
I believe the object of this issue was to keep the bash files up to date, but this PR only adds a test to check for the file, or optionally update it. Is there something that can be done to make sure that the file is updated or generated on every compile, perhaps by modifying the Makefile to run this test everytime?  In that case, I believe that the "--overwrite" option should be the default.

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -290,8 +290,7 @@ static RPCHelpMan addnode()
     if (!request.params[1].isNull())
         strCommand = request.params[1].get_str();
     if (strCommand != "onetry" && strCommand != "add" && strCommand != "remove") {
-        throw std::runtime_error(
-            self.ToString());
+        throw std::runtime_error(self.ToString());
```
Comment (pablomartin4btc) [inline src/rpc/net.cpp:293]: Due to the amount of files changed and the context of this PR, perhaps we could leave this out(?) and follow it up in a different change.

**Comment by achow101** [comment]
Closed and reopened to hopefully trigger CI

**Comment by suhailsaqan** [comment]
@achow101 Any idea why CI is not running on this PR?

**Comment by fanquake** [comment]
> achow101 Any idea why CI is not running on this PR?

Rebasing may help. While you are at it, you can also squash your commits.

**Comment by achow101** [comment]
Are you still working on this?

**Pull request closed without merging**